### PR TITLE
Automated upgrade of multi-workers with no extra node

### PR DESCRIPTION
**Pull request by Xenwar**
Boot disk image upgrade for multi-worker cluster with no extra node.

### CODE REVIEW

**Comment by Jaakko-Os** [comment]
Looks good.
Refactoring and polishing to be done in separate PR as discussed.

**Comment by smoshiur1237** [comment]
Looks good to me

File: upgrade_tests/common.sh [hunk 1]
```diff
@@ -144,3 +144,46 @@ function wait_for_orig_node_deprovisioned() {
     fi
     done
 }
+
+
+function wait_for_ctrlplane_to_scale_out {
+    ORIGINAL_NODE="${1}"
+    echo "Scaling out controlplane nodes"
+    kubectl get kcp -n metal3 test1 -o json | jq '.spec.replicas=3' | kubectl apply -f-
```
Comment (kashifest) [inline upgrade_tests/common.sh:152]: parameterizing the namespace would be ideal

File: upgrade_tests/common.sh [hunk 2]
```diff
@@ -144,3 +144,46 @@ function wait_for_orig_node_deprovisioned() {
...
+            echo "Error: Scaling out of controlplane nodes took to long"
+            exit 1
+        fi        
+    done
+}
+
+function wait_for_worker_to_scale_out {
+    ORIGINAL_NODE="${1}"
+    echo "Scaling out worker nodes"    
+    kubectl get machinedeployment -n metal3 test1 -o json | jq '.spec.replicas=3' | kubectl apply -f-
```
Comment (kashifest) [inline upgrade_tests/common.sh:173]: parameterizing the namespace and name would be ideal

File: upgrade_tests/workers_upgrade/w_multi-worker-no-extra-node_ugprade.sh [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+wait_for_ctrlplane_provisioning_complete ${CP_NODE} ${CP_NODE_IP} "original controlplane node"
+# apply CNI
+ssh -o PasswordAuthentication=no -o "StrictHostKeyChecking no" "${UPGRADE_USER}@${CP_NODE_IP}" -- kubectl apply -f https://docs.projectcalico.org/manifests/calico.yaml
+provision_worker_node
+
+wait_for_worker_to_scale_out ${CP_NODE_IP}
+
+echo "Create a new metal3MachineTemplate with new boot disk image for worker nodes"
+wr_Metal3MachineTemplate_OUTPUT_FILE="/tmp/wr_new_image.yaml"
+CLUSTER_UID=$(kubectl get clusters -n metal3 test1 -o json |jq '.metadata.uid' | cut -f2 -d\")
```
Comment (kashifest) [inline upgrade_tests/workers_upgrade/w_multi-worker-no-extra-node_ugprade.sh:26]: parameterizing the namespace and name would be ideal

File: upgrade_tests/workers_upgrade/w_multi-worker-no-extra-node_ugprade.sh [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+generate_metal3MachineTemplate "${new_wr_metal3MachineTemplate_name}" "${CLUSTER_UID}" "${wr_Metal3MachineTemplate_OUTPUT_FILE}"
+
+kubectl apply -f "${wr_Metal3MachineTemplate_OUTPUT_FILE}"
+
+# Change metal3MachineTemplate references.
+kubectl get machinedeployment -n metal3 test1 -o json | jq '.spec.strategy.rollingUpdate.maxSurge=1|.spec.strategy.rollingUpdate.maxUnavailable=1' | kubectl apply -f-
+kubectl get machinedeployment -n metal3 test1 -o json | jq '.spec.template.spec.infrastructureRef.name="test1-new-workers-image"' | kubectl apply -f-
```
Comment (kashifest) [inline upgrade_tests/workers_upgrade/w_multi-worker-no-extra-node_ugprade.sh:33]: parameterizing the namespace and name would be ideal

File: upgrade_tests/workers_upgrade/w_multi-worker-no-extra-node_ugprade.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+
+echo "Waiting for start of boot disk upgrade of all worker nodes"
+for i in {1..3600};do
+  count=$(kubectl get bmh -n metal3 | awk 'NR>1'| grep -i ${new_wr_metal3MachineTemplate_name} | wc -l)
```
Comment (kashifest) [inline upgrade_tests/workers_upgrade/w_multi-worker-no-extra-node_ugprade.sh:38]: same

**Pull request merged**